### PR TITLE
Support @AutoDefault for 'required-key-but-not-required-value' behavior

### PR DESCRIPTION
**Pull request by sebastiananu**
Implemented the new annotation @AutoDefault to support, assigning a default value when the value of the property is set to 'auto'.
Related issue:https://github.com/brianm/config-magic/issues/15

### CODE REVIEW

**Comment by brianm** [comment]
It has been a while since I have worked on this library, so it is probably something I don't remember, but..  how is `@AutoDefault` different from `@Default` ?

**Comment by brianm** [comment]
Oh, I think I get it, iff the value is "auto" it will ignore the value and use the default.

It might be cleaner to implement as another attribute on `@Default`, something like `@Default(value="kangaroo", overrides={"auto", "défaut", "default"})` where the values in the `overrides` attribute are values the default value should override. There might be a better name for `overrides` though.

**Comment by divyekapoor** [comment]
@brianm There are 2 challenges with the suggested approach:
1. If `@Default` picks up a second parameter, it will force a code change on all existing code: `@Default("foo")` will not work anymore. The compiler will ask for `@Default(value = "foo")` and that's probably not desired at all.
2. The behavior of `@Default` is different from `@AutoDefault` (`@Default` allows an absent key; `@AutoDefault` does not). 
3. The closest we can get to addressing your comment is to make `@AutoDefault` more verbose. ie. Change:
`@AutoDefault("value")` to `@AutoDefault(value = "value", for_values = {"auto", "default", "AUTO"})` - I'm not sure that's desirable, but it will work.

Couple of things that would do with your input:
1. We would prefer clarity & simplicity: `@AutoDefault` should override "auto" / "Auto" / "AUTO" - it's also in line with `@Default`. It's also follows YAGNI and is the smallest 4 chars (fewer chances of a typo). 
2. If you would strongly prefer that the override string be configurable, then the proposal is: `@AutoDefault(value = "foo", for_values = {"auto", "AUTO", "Auto", "default"})`

Please confirm whether we should proceed with (1) or (2). I would (personally) prefer (1) but will be open to (2) if that's your preference. 

Thanks for working with us!

**Comment by brianm** [comment]
@divyekapoor Re: various points

(1): If the new param for @Default has a default value (empty array), I *believe* it is binary backwards compatible (this is easy to test)

(2) We can get around that by another Default param, "required" which defaults to false.

I think having both Default/AutoDefault is harder to use as there are no affordances for which a user wants, just RTFM. I lean towards modifying `@Default` as this really is having a default... with slightly different (parameterized) behavior. I think the behavior you want is reasonable, so want to do it in the simplest way :-)

**Comment by divyekapoor** [comment]
Thanks for the feedback! 

@sebastiananu just tested this - `@Default("foo")` works when a 2nd parameter is added with a default. 
The updated proposal is:
`@Default("foo")` - works as before (implicit 2nd param's default value is "*") - key is not required.
`@Default(value = "foo", for_values = {"auto", "AUTO"})` - is the new addition - key is required and only if value is set to "auto" or "AUTO"  will it get substituted. 

Multiple `@Default` with different for_values will not be supported (only a single supported annotation). 

Please confirm (especially the last part) and we can get going. 

Cheers!

**Comment by brianm** [comment]
I would default it to the empty set, not a flag value ("*"). There is an implicit "null is also replaced by default value" when property is not present, but null as special case I think is fine. It is a pain.

Nit, I'd stick with camelCase instead of `for_values` as much as I prefer snake_case.

**Comment by brianm** [comment]
BTW, happy to confirm that there should not be multiple `@Default` annotations, that would end to horrible behavior. We should probably eagerly raise an exception (IllegalState) if it is detected!

File: src/main/java/org/skife/config/Default.java [hunk 1]
```diff
@@ -10,4 +10,5 @@
 public @interface Default
 {
     String value();
+    String[] for_values() default "*";
```
Comment (divyekapoor) [inline src/main/java/org/skife/config/Default.java:13]: Make this "" so that it will never match

Comment (divyekapoor) [reply src/main/java/org/skife/config/Default.java:13]: Or make this null.

Comment (divyekapoor) [reply src/main/java/org/skife/config/Default.java:13]: Please also verify with a test what for_values returns when the default is used (eg. as in `@Default("foo")`)

File: src/test/java/org/skife/config/TestDefaultsPresent.java [hunk 2]
```diff
@@ -54,6 +54,27 @@ public void testAbstractClassDefaultNull()
...
+    
+    @Test
+    public void testClassDefaultForValues()
+    {
+    	Properties p = new Properties();
+    	p.setProperty("value", "auto");
+    	cof = new ConfigurationObjectFactory(p);
+    	EmptyForValueClass ec = cof.build(EmptyForValueClass.class);
+
+        Assert.assertEquals("auto-value", ec.getValue());
```
Comment (divyekapoor) [inline src/test/java/org/skife/config/TestDefaultsPresent.java:66]: Please verify assignedFrom in this test.

**Review by divyekapoor — APPROVED** [review]
This PR looks good to me. Please make the minor nit changes. Thanks!

File: src/test/java/org/skife/config/TestDefaultsPresent.java [hunk 3]
```diff
@@ -54,6 +54,36 @@ public void testAbstractClassDefaultNull()
...
+    public void testClassDefaultForValues()
+    {
+    	Properties p = new Properties();
+    	p.setProperty("value", "auto");
+    	cof = new ConfigurationObjectFactory(p);
+    	EmptyForValueClass ec = cof.build(EmptyForValueClass.class);
+        Assert.assertEquals("auto-value", ec.getValue());
+    }
+    
+    @Test
```
Comment (brianm) [inline src/test/java/org/skife/config/TestDefaultsPresent.java:68]: Could you add a negative test?

File: src/main/java/org/skife/config/Default.java [hunk 4]
```diff
@@ -10,4 +10,5 @@
 public @interface Default
 {
     String value();
+    String[] for_values() default {};
```
Comment (brianm) [inline src/main/java/org/skife/config/Default.java:13]: I like `snake_case`, but I *think* `camelCase` is still idiomatic naming for attribute values. Any reason to maintain snake_case here?

File: src/main/java/org/skife/config/ConfigurationObjectFactory.java [hunk 5]
```diff
@@ -188,6 +194,7 @@ private void buildSimple(List<Callback> callbacks, Method method, Config annotat
         //   ignore the passed in value (which will be null)
         // - if all else fails, throw an exception.
```
Comment (brianm) [inline src/main/java/org/skife/config/ConfigurationObjectFactory.java:195]: Could you expand this explanatory comment to make accurate for this feature? The code is hairy enough that I would like to the keep the explanation in sync.

**Review by brianm — CHANGES_REQUESTED** [review]

**Review by brianm — APPROVED** [review]

**Comment by brianm** [comment]
Merged, am cutting release, will be 0.19

**Comment by brianm** [comment]
Hmm, something failed with the oss.s.o release, need to debug into what happened.

**Comment by sebastiananu** [comment]
Thanks, Brianm for approving the pull request. This helps a lot. 
Also, Just would like to double-check, have you released this version?

**Comment by brianm** [comment]
Yep, I checked and it looks like it made it to central. [0.22 should have these changes](https://github.com/brianm/config-magic/blob/config-magic-0.22/src/test/java/org/skife/config/TestDefaultsPresent.java#L88-L96)

**Pull request merged**